system/include/emscripten/dom_pk_codes.h

```diff
@@ -0,0 +1,506 @@
+#define DOM_PK_EJECT                0xE02C /* "Eject"              */
+#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "AudioVolumeDown"    */
+#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "VolumeDown"         */
+#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "AudioVolumeUp"      */
+#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "VolumeUp"           */
+#define DOM_PK_BROWSER_HOME         0xE032 /* "BrowserHome"        */
+#define DOM_PK_NUMPAD_DIVIDE        0xE035 /* "NumpadDivide"       */
+#define DOM_PK_ALT_RIGHT            0xE038 /* "AltRight"           */
+#define DOM_PK_HELP                 0xE03B /* "Help"               */
+#define DOM_PK_NUM_LOCK             0xE045 /* "NumLock"            */
+#define DOM_PK_HOME                 0xE047 /* "Home"               */
+#define DOM_PK_ARROW_UP             0xE048 /* "ArrowUp"            */
+#define DOM_PK_PAGE_UP              0xE049 /* "PageUp"             */
+#define DOM_PK_ARROW_LEFT           0xE04B /* "ArrowLeft"          */
+#define DOM_PK_ARROW_RIGHT          0xE04D /* "ArrowRight"         */
+#define DOM_PK_END                  0xE04F /* "End"                */
+#define DOM_PK_ARROW_DOWN           0xE050 /* "ArrowDown"          */
+#define DOM_PK_PAGE_DOWN            0xE051 /* "PageDown"           */
+#define DOM_PK_INSERT               0xE052 /* "Insert"             */
+#define DOM_PK_DELETE               0xE053 /* "Delete"             */
+#define DOM_PK_META_LEFT            0xE05B /* "MetaLeft"           */
+#define DOM_PK_OS_LEFT              0xE05B /* "OSLeft"             */
+#define DOM_PK_META_RIGHT           0xE05C /* "MetaRight"          */
+#define DOM_PK_OS_RIGHT             0xE05C /* "OSRight"            */
+#define DOM_PK_CONTEXT_MENU         0xE05D /* "ContextMenu"        */
+#define DOM_PK_POWER                0xE05E /* "Power"              */
+#define DOM_PK_BROWSER_SEARCH       0xE065 /* "BrowserSearch"      */
+#define DOM_PK_BROWSER_FAVORITES    0xE066 /* "BrowserFavorites"   */
+#define DOM_PK_BROWSER_REFRESH      0xE067 /* "BrowserRefresh"     */
+#define DOM_PK_BROWSER_STOP         0xE068 /* "BrowserStop"        */
+#define DOM_PK_BROWSER_FORWARD      0xE069 /* "BrowserForward"     */
+#define DOM_PK_BROWSER_BACK         0xE06A /* "BrowserBack"        */
+#define DOM_PK_LAUNCH_APP_1         0xE06B /* "LaunchApp1"         */
+#define DOM_PK_LAUNCH_MAIL          0xE06C /* "LaunchMail"         */
+#define DOM_PK_LAUNCH_MEDIA_PLAYER  0xE06D /* "LaunchMediaPlayer"  */
+#define DOM_PK_MEDIA_SELECT         0xE06D /* "MediaSelect"        */
+
+/* Maps the EmscriptenKeyboardEvent::code field from emscripten/html5.h to one of the DOM_PK codes above. */
+static DOM_PK_CODE_TYPE emscripten_compute_dom_pk_code(const char *keyCodeString)
+{
+  if (!keyCodeString) return 0;
+
+  /* Compute the collision free hash. */
+  unsigned int hash = 0;
+  while(*keyCodeString) hash = ((hash ^ 0x5EE148E8U) << 3) ^ (unsigned int)*keyCodeString++;
```

Comment:
> One could switch-case on first few characters and then fall back to linear strcmp for the remaining ones, but that seems much more complex to read and debug than this

I guess this comes down to a matter of opinion, but for me the hashing is harder to read and debug. On the other hand, I agree it's not that hard. So I'm ok either way, but maybe we should hear other opinions on this.

I worry more about the risk of not rejecting invalid inputs. But I think I may have misunderstood the code before, is `emscripten_compute_dom_pk_code` not going to be called by users, but just internally? That seems ok then.